lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3419,6 +3430,7 @@ void AdjointEmitter::materializeZeroIndirect(CanType type,
                                              SILValue bufferAccess,
                                              SILLocation loc) {
   // Lookup `AdditiveArithmetic.zero.getter`.
+  type = remapASTType(type);
```

Comment:
`materializeZeroIndirect` should in principle assume `type` has been processed. Did only substituting in visitors not solve your problem?

Reply:
Ah, I came from a different perspective, which is to put the type remapping logic in the `materializeZero` functions to lessen the burden of putting it in all the visitors.

I also see your perspective: `materializeZero` functions should care only about "materializing zero", not remapping types.

What do you think?

Reply:
Neither seems to be the best approach.  There's another way of thinking: `materializeAdjoint` is always about materializing `AdjointValue`s. The simplest approach should be to transform the type when creating an `AdjointValue` (in its c'tor and static factory methods). This makes sure all visitors will get the values of the right type because they only call `getAdjointValue`.